src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -0,0 +1,415 @@
+#include <util/expr_iterator.h>
+#include <util/string_utils.h>
+
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+
+namespace
+{
+void typecheck_function_pointer_restrictions(
+  const goto_modelt &goto_model,
+  const function_pointer_restrictionst &restrictions)
+{
+  for(auto const &restriction : restrictions.restrictions)
+  {
+    auto const function_pointer_sym =
+      goto_model.symbol_table.lookup(restriction.first);
+    if(function_pointer_sym == nullptr)
+    {
+      throw invalid_command_line_argument_exceptiont{
+        id2string(restriction.first) + " not found in the symbol table",
+        "--restrict-function-pointer"};
+    }
+    auto const &function_pointer_type = function_pointer_sym->type;
+    if(function_pointer_type.id() != ID_pointer)
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "not a function pointer: " + id2string(restriction.first),
+        "--restrict-function-pointer"};
+    }
+    auto const &function_type =
+      to_pointer_type(function_pointer_type).subtype();
+    if(function_type.id() != ID_code)
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "not a function pointer: " + id2string(restriction.first),
+        "--restrict-function-pointer"};
+    }
+    auto const &ns = namespacet{goto_model.symbol_table};
+    for(auto const &function_pointer_target : restriction.second)
+    {
+      auto const function_pointer_target_sym =
+        goto_model.symbol_table.lookup(function_pointer_target);
+      if(function_pointer_target_sym == nullptr)
+      {
+        throw invalid_command_line_argument_exceptiont{
+          "symbol not found: " + id2string(function_pointer_target),
+          "--restrict-function-pointer"};
+      }
+      auto const &function_pointer_target_type =
+        function_pointer_target_sym->type;
+      if(function_type != function_pointer_target_type)
+      {
+        throw invalid_command_line_argument_exceptiont{
+          "type mismatch: `" + id2string(restriction.first) + "' points to `" +
+            type2c(function_type, ns) + "', but restriction `" +
+            id2string(function_pointer_target) + "' has type `" +
+            type2c(function_pointer_target_type, ns) + "'",
+          "--restrict-function-pointer"};
+      }
+    }
+  }
+}
+
+source_locationt make_function_pointer_restriction_assertion_source_location(
+  source_locationt source_location,
+  const function_pointer_restrictionst::value_type restriction)
+{
+  std::stringstream comment;
+  comment << "dereferenced function pointer at " << restriction.first
+          << " must be ";
+  if(restriction.second.size() == 1)
+  {
+    comment << *restriction.second.begin();
+  }
+  else
+  {
+    comment << "one of [";
+    bool first = true;
+    for(auto const &target : restriction.second)
+    {
+      if(!first)
+      {
+        comment << ", ";
+      }
+      else
+      {
+        first = false;
+      }
+      comment << target;
+    }
```

Comment:
:idea: `join_strings` handles this for you
```suggestion
    join_strings(comment, restrictions.second.begin(), restrictions.second.end(), ',');
```

Reply:
I didn’t know we had that